Move format_args!() into AST (and expand it during AST lowering)

m-ou-se opened the pull request:
Part of #99012

Implements https://github.com/rust-lang/compiler-team/issues/541

This moves FormatArgs from rustc_builtin_macros to rustc_ast_lowering. For now, the end result is the same. But this allows for future changes to do smarter things with format_args!(). It also allows Clippy to directly access the ast::FormatArgs, making things a lot easier.

This change turns the format args types into lang items. The builtin macro used to refer to them by their path. After this change, the path is no longer relevant, making it easier to make changes in `core`.

This updates clippy to use the new language items, but this doesn't yet make clippy use the ast::FormatArgs structure that's now available. That should be done after this is merged.

m-ou-se committed:
Add some helper functions to LoweringContext.

m-ou-se committed:
Turn format arguments types into lang items.

m-ou-se committed:
Add new fn to core::fmt::rt::v1::Argument.

m-ou-se committed:
Expand format_args!() in rust_ast_lowering.

m-ou-se committed:
Expose some LoweringContext methods.

m-ou-se committed:
Bless tests.

m-ou-se committed:
Make clippy compile.

m-ou-se commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

m-ou-se committed:
Update rustfmt for ast::ExprKind::FormatArgs.

Rustfmt doesn't expand macros, so that's easy: FormatArgs nodes do not
occur in the unexpanded AST.

m-ou-se committed:
Bless pretty tests.

m-ou-se committed:
Update clippy for new format_args!() lang items.

m-ou-se commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 36fe3b971dbf23bbb362f79006f7c8571818eb5d with merge bbf793deb381ce142c8b65dba5caf69e66a71f55...
<!-- homu: {"type":"TryBuildStarted","head_sha":"36fe3b971dbf23bbb362f79006f7c8571818eb5d","merge_sha":"bbf793deb381ce142c8b65dba5caf69e66a71f55"} -->

m-ou-se committed:
Make core::fmt::rt::v1::Argument::new const+inline.

m-ou-se committed:
Update outdated comment.

m-ou-se committed:
Update comment explaining format_args!() expansion.

m-ou-se commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 56593104acef6afc0f3d29c765801bb44ec02c65 with merge e36affffe97378a0027b4bcfbb18d27356164ed0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"56593104acef6afc0f3d29c765801bb44ec02c65","merge_sha":"e36affffe97378a0027b4bcfbb18d27356164ed0"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3900083144/jobs/6662257307)
Build commit: e36affffe97378a0027b4bcfbb18d27356164ed0 (`e36affffe97378a0027b4bcfbb18d27356164ed0`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3900083144/jobs/6662257307"},"merge_sha":"e36affffe97378a0027b4bcfbb18d27356164ed0"} -->

rust-timer commented:
Queued e36affffe97378a0027b4bcfbb18d27356164ed0 with parent d4203eda5f0dd0b0192bfd22ea5989fbaf917818, future [comparison URL](https://perf.rust-lang.org/compare.html?start=d4203eda5f0dd0b0192bfd22ea5989fbaf917818&end=e36affffe97378a0027b4bcfbb18d27356164ed0).
<!-- rust-timer: temporary -->

oli-obk commented on compiler/rustc_ast_lowering/src/format.rs:
```diff
@@ -0,0 +1,354 @@
+    arg: &'hir hir::Expr<'hir>,
+    ty: ArgumentType,
+) -> hir::Expr<'hir> {
+    // Generate:
+    //     ::core::fmt::ArgumentV1::new_…(arg)
```
nit: all of these comments could be doc comments on the functions. We do generate docs for private items in rustc.

oli-obk commented on compiler/rustc_ast_lowering/src/format.rs:
```diff
@@ -0,0 +1,354 @@
+                let FormatArgsPiece::Placeholder(placeholder) = piece else { return None };
+                Some(make_format_spec(ctx, macsp, placeholder, &mut argmap))
+            })
+            .collect();
```
this `collect` is unnecessary.

m-ou-se replied:
It is necessary. The closure and therefore the iterator mutably borrows `ctx`, so we can't pass that iterator to `ctx.arena`.

oli-obk commented on compiler/rustc_ast_lowering/src/format.rs:
```diff
@@ -0,0 +1,354 @@
+            }
+        }
+
+        fn visit_mac_call(&mut self, _: &ast::MacCall) {
+            self.0 = true;
+        }
```
This should be unreachable at this point, all macro expansion must be done before ast lowering

m-ou-se replied:
Ah, right. I just moved this from rustc_builtin_macros/src/format/expand.rs, without realising that some parts are no longer relevant.

oli-obk commented on compiler/rustc_ast_lowering/src/format.rs:
```diff
@@ -0,0 +1,354 @@
+            self.0 = true;
+        }
+
+        fn visit_attribute(&mut self, _: &ast::Attribute) {
+            // Conservatively assume this may be a proc macro attribute in
+            // expression position.
+            self.0 = true;
+        }
```
This should never be an issue, all such attributes ~~should~~ must already have been expanded.

oli-obk commented on compiler/rustc_ast_lowering/src/format.rs:
```diff
@@ -0,0 +1,354 @@
+        ctx.expr_array_ref(macsp, ctx.arena.alloc_from_iter(elements))
+    } else {
+        // Generate:
+        //     &match (&arg0, &arg1, &arg2) {
```
```suggestion
        //     &match (&arg0, &arg1) {
```

oli-obk commented on compiler/rustc_ast_lowering/src/format.rs:
```diff
@@ -0,0 +1,354 @@
+
+    let arguments = fmt.arguments.all_args();
+
+    // If the args array contains exactly all the original arguments once,
+    // in order, we can use a simple array instead of a `match` construction.
```
Please document why we do this optimization at all, instead of always using the `match` version

m-ou-se replied:
Hm, it might just have been to keep the `-Zunpretty=expanded` output less verbose. I'll go dig a bit through the git and issue history to find out.

m-ou-se replied:
Yes, this is the original PR: https://github.com/rust-lang/rust/pull/91359

I think we care less about a verbose `-Zunpretty=hir` than `-Zunpretty=expanded`, so I'll just remove this.

m-ou-se replied:
Oh actually, that PR did result in a lot of performance improvements: https://perf.rust-lang.org/compare.html?start=523be2e05da322daaecf1ecc8f2c0d625f5f46e3&end=0bcacb391b28460f5a50fd627f01f670dfcfc7cc

So let's run another perf test with this case removed, to see if it makes the same difference here at this level.

m-ou-se replied:
Here are the perf results, comparing this PR with the 'simple array' special case, to a version of this PR with that special case removed: https://perf.rust-lang.org/compare.html?start=e36affffe97378a0027b4bcfbb18d27356164ed0&end=5dbee4d3a6728eb4530fb66c9775834438ecec74&stat=instructions:u

Unfortunately, all red numbers. :(

m-ou-se replied:
Added a comment about this with a link to these perf results.

oli-obk commented on compiler/rustc_ast_lowering/src/format.rs:
```diff
@@ -0,0 +1,354 @@
+                    hir::ExprKind::AddrOf(hir::BorrowKind::Ref, hir::Mutability::Not, arg_expr),
+                )
+            })
+            .collect();
```
unnecessary `collect`

m-ou-se replied:
Same as above: The closure and therefore the iterator mutably borrows `ctx`, so we can't pass that iterator to `ctx.arena`.

rust-timer commented:
Finished benchmarking commit (e36affffe97378a0027b4bcfbb18d27356164ed0): [comparison URL](https://perf.rust-lang.org/compare.html?start=d4203eda5f0dd0b0192bfd22ea5989fbaf917818&end=e36affffe97378a0027b4bcfbb18d27356164ed0&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=d4203eda5f0dd0b0192bfd22ea5989fbaf917818&end=e36affffe97378a0027b4bcfbb18d27356164ed0&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.8%  | [0.8%, 0.8%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.5%  | [0.2%, 1.5%]   | 6     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-1.4%, -0.2%] | 47    |
| Improvements ✅ <br /> (secondary) | -0.4% | [-0.6%, -0.3%] | 4     |
| All ❌✅ (primary)                 | -0.6% | [-1.4%, 0.8%]  | 48    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d4203eda5f0dd0b0192bfd22ea5989fbaf917818&end=e36affffe97378a0027b4bcfbb18d27356164ed0&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.7%  | [4.7%, 4.7%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.8%  | [1.0%, 7.9%]   | 7     |
| Improvements ✅ <br /> (primary)   | -3.2% | [-5.2%, -0.5%] | 7     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-2.7%, -1.4%] | 7     |
| All ❌✅ (primary)                 | -2.3% | [-5.2%, 4.7%]  | 8     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d4203eda5f0dd0b0192bfd22ea5989fbaf917818&end=e36affffe97378a0027b4bcfbb18d27356164ed0&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.8% | [-2.9%, -0.9%] | 5     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.8% | [-2.9%, -0.9%] | 5     |
</details>

petrochenkov commented on compiler/rustc_ast_pretty/src/pprust/state/expr.rs:
```diff
@@ -528,6 +530,18 @@ impl<'a> State<'a> {
                 self.print_inline_asm(a);
             }
+            ast::ExprKind::FormatArgs(fmt) => {
+                self.word("format_args!");
```
This is not correct, as discussed in the zulip thread, neither `asm!` nor `format_args!` should do this, but should use some canonical expanded token-based representation instead.
If you don't plan to implement it in this PR (I admit it's not super urgent), then it needs at leas a FIXME and comment.

m-ou-se replied:
Ah, right. Expanding `std::arch::asm!()` now results in just `asm!()`, which is a bit odd. Maybe it should be something like `builtin#asm()` and `builtin#format_args()`, but that probably needs some bikeshedding.

rust-log-analyzer commented:
The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3896627108/jobs/6653394282) [(plain)](https://github.com/rust-lang-ci/rust/commit/225ab3814cb608b5f14923fe372d5713e79a2b16/checks/10589170389/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[2023-01-11T21:52:35Z DEBUG collector::execute] Benchmark iteration 1/1
[2023-01-11T21:52:36Z INFO  collector::execute] run_rustc with incremental=false, profile=Check, scenario=Some(Full), patch=None
[2023-01-11T21:52:36Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpki6hxM#cargo@0.60.0" "--profile" "check" "--lib" "--" "--wrap-rustc-with" "Eprintln"
[2023-01-11T21:53:00Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrFull), patch=None
[2023-01-11T21:53:00Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpki6hxM#cargo@0.60.0" "--profile" "check" "--lib" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpki6hxM/incremental-state"
[2023-01-11T21:53:30Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrUnchanged), patch=None
[2023-01-11T21:53:30Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpki6hxM#cargo@0.60.0" "--profile" "check" "--lib" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpki6hxM/incremental-state"
[2023-01-11T21:53:36Z DEBUG collector::execute] applying println to "/tmp/.tmpki6hxM"
[2023-01-11T21:53:36Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2023-01-11T21:53:36Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2023-01-11T21:53:36Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpki6hxM#cargo@0.60.0" "--profile" "check" "--lib" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpki6hxM/incremental-state"
[2023-01-11T21:53:42Z DEBUG collector::execute] Benchmark iteration 1/1
[2023-01-11T21:53:42Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=Some(Full), patch=None
[2023-01-11T21:53:42Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp3P9uHK#cargo@0.60.0" "--lib" "--" "--wrap-rustc-with" "Eprintln"
[2023-01-11T21:54:46Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None
---
[2023-01-11T21:56:31Z DEBUG collector::execute] Benchmark iteration 1/1
[2023-01-11T21:56:32Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None
[2023-01-11T21:56:32Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpjwRQyE#cargo@0.60.0" "--release" "--lib" "--" "--wrap-rustc-with" "Eprintln"
[2023-01-11T21:57:49Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
[2023-01-11T21:57:49Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpjwRQyE#cargo@0.60.0" "--release" "--lib" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpjwRQyE/incremental-state"
[2023-01-11T21:59:07Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrUnchanged), patch=None
[2023-01-11T21:59:07Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpjwRQyE#cargo@0.60.0" "--release" "--lib" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpjwRQyE/incremental-state"
[2023-01-11T21:59:22Z DEBUG collector::execute] applying println to "/tmp/.tmpjwRQyE"
[2023-01-11T21:59:22Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2023-01-11T21:59:22Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2023-01-11T21:59:22Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpjwRQyE#cargo@0.60.0" "--release" "--lib" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpjwRQyE/incremental-state"
Executing benchmark ctfe-stress-5 (3/8)
Preparing ctfe-stress-5
[2023-01-11T21:59:36Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=None, patch=None
[2023-01-11T21:59:36Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=None, patch=None
---
[2023-01-11T22:01:07Z DEBUG collector::execute] Benchmark iteration 1/1
[2023-01-11T22:01:07Z INFO  collector::execute] run_rustc with incremental=false, profile=Check, scenario=Some(Full), patch=None
[2023-01-11T22:01:07Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpxGkLST#diesel@1.4.8" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln"
[2023-01-11T22:01:26Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrFull), patch=None
[2023-01-11T22:01:26Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpxGkLST#diesel@1.4.8" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpxGkLST/incremental-state"
[2023-01-11T22:01:51Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrUnchanged), patch=None
[2023-01-11T22:01:51Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpxGkLST#diesel@1.4.8" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpxGkLST/incremental-state"
[2023-01-11T22:01:55Z DEBUG collector::execute] applying println to "/tmp/.tmpxGkLST"
[2023-01-11T22:01:55Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2023-01-11T22:01:55Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2023-01-11T22:01:55Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpxGkLST#diesel@1.4.8" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpxGkLST/incremental-state"
[2023-01-11T22:01:59Z DEBUG collector::execute] Benchmark iteration 1/1
[2023-01-11T22:02:00Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=Some(Full), patch=None
[2023-01-11T22:02:00Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpY9U2Kf#diesel@1.4.8" "--" "--wrap-rustc-with" "Eprintln"
[2023-01-11T22:02:24Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None
---
[2023-01-11T22:03:06Z DEBUG collector::execute] Benchmark iteration 1/1
[2023-01-11T22:03:06Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None
[2023-01-11T22:03:06Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpOAxrWs#diesel@1.4.8" "--release" "--" "--wrap-rustc-with" "Eprintln"
[2023-01-11T22:03:30Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
[2023-01-11T22:03:30Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpOAxrWs#diesel@1.4.8" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpOAxrWs/incremental-state"
[2023-01-11T22:04:03Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrUnchanged), patch=None
[2023-01-11T22:04:03Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpOAxrWs#diesel@1.4.8" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpOAxrWs/incremental-state"
[2023-01-11T22:04:08Z DEBUG collector::execute] applying println to "/tmp/.tmpOAxrWs"
[2023-01-11T22:04:08Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2023-01-11T22:04:08Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2023-01-11T22:04:08Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpOAxrWs#diesel@1.4.8" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpOAxrWs/incremental-state"
Executing benchmark externs (5/8)
Preparing externs
[2023-01-11T22:04:13Z INFO  collector::execute] run_rustc with incremental=false, profile=Check, scenario=None, patch=None
[2023-01-11T22:04:13Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=None, patch=None
---
[2023-01-11T22:04:19Z DEBUG collector::execute] Benchmark iteration 1/1
[2023-01-11T22:04:19Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None
[2023-01-11T22:04:19Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpFZzHHy#externs@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln"
[2023-01-11T22:04:20Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
[2023-01-11T22:04:20Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpFZzHHy#externs@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpFZzHHy/incremental-state"
[2023-01-11T22:04:21Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrUnchanged), patch=None
[2023-01-11T22:04:21Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpFZzHHy#externs@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpFZzHHy/incremental-state"
Executing benchmark match-stress (6/8)
Preparing match-stress
[2023-01-11T22:04:22Z INFO  collector::execute] run_rustc with incremental=false, profile=Check, scenario=None, patch=None
[2023-01-11T22:04:22Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=None, patch=None
---
[2023-01-11T22:04:22Z DEBUG collector::execute] Benchmark iteration 1/1
[2023-01-11T22:04:22Z INFO  collector::execute] run_rustc with incremental=false, profile=Check, scenario=Some(Full), patch=None
[2023-01-11T22:04:22Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpcyXTIm#match-stress@0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln"
[2023-01-11T22:04:26Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrFull), patch=None
[2023-01-11T22:04:26Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpcyXTIm#match-stress@0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpcyXTIm/incremental-state"
[2023-01-11T22:04:29Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrUnchanged), patch=None
[2023-01-11T22:04:29Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpcyXTIm#match-stress@0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpcyXTIm/incremental-state"
[2023-01-11T22:04:31Z DEBUG collector::execute] Benchmark iteration 1/1
[2023-01-11T22:04:31Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=Some(Full), patch=None
[2023-01-11T22:04:31Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpzPGKyA#match-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln"
[2023-01-11T22:04:34Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None
[2023-01-11T22:04:34Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None
[2023-01-11T22:04:34Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpzPGKyA#match-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpzPGKyA/incremental-state"
[2023-01-11T22:04:39Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrUnchanged), patch=None
[2023-01-11T22:04:39Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpzPGKyA#match-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpzPGKyA/incremental-state"
[2023-01-11T22:04:40Z DEBUG collector::execute] Benchmark iteration 1/1
[2023-01-11T22:04:40Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None
[2023-01-11T22:04:40Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpPRVpY3#match-stress@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln"
[2023-01-11T22:04:44Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
[2023-01-11T22:04:44Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
[2023-01-11T22:04:44Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpPRVpY3#match-stress@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpPRVpY3/incremental-state"
[2023-01-11T22:04:48Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrUnchanged), patch=None
[2023-01-11T22:04:48Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpPRVpY3#match-stress@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpPRVpY3/incremental-state"
Executing benchmark token-stream-stress (7/8)
Preparing token-stream-stress
[2023-01-11T22:04:49Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=None, patch=None
[2023-01-11T22:04:49Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=None, patch=None
---
[2023-01-11T22:04:54Z DEBUG collector::execute] Benchmark iteration 1/1
[2023-01-11T22:04:54Z INFO  collector::execute] run_rustc with incremental=false, profile=Check, scenario=Some(Full), patch=None
[2023-01-11T22:04:54Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpXiccdE#tuple-stress@0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln"
[2023-01-11T22:04:59Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrFull), patch=None
[2023-01-11T22:04:59Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpXiccdE#tuple-stress@0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpXiccdE/incremental-state"
[2023-01-11T22:05:06Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrUnchanged), patch=None
[2023-01-11T22:05:06Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpXiccdE#tuple-stress@0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpXiccdE/incremental-state"
[2023-01-11T22:05:07Z DEBUG collector::execute] applying new row to "/tmp/.tmpXiccdE"
[2023-01-11T22:05:08Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("new row"), path: "0-new-row.patch" })
[2023-01-11T22:05:08Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("new row"), path: "0-new-row.patch" })
[2023-01-11T22:05:08Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpXiccdE#tuple-stress@0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpXiccdE/incremental-state"
[2023-01-11T22:05:14Z DEBUG collector::execute] Benchmark iteration 1/1
[2023-01-11T22:05:14Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=Some(Full), patch=None
[2023-01-11T22:05:14Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp1xumuB#tuple-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln"
[2023-01-11T22:05:20Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None
---
[RUSTC-TIMING] curl_sys test:false 0.105
    Checking rustc-workspace-hack v1.0.0 (/checkout/src/tools/rustc-workspace-hack)
[RUSTC-TIMING] rustc_workspace_hack test:false 0.033
 Documenting rustfmt-nightly v1.5.1 (/checkout/src/tools/rustfmt)
error[E0004]: non-exhaustive patterns: `ExprKind::FormatArgs(_)` not covered
    |
463 |     match expr.kind {
463 |     match expr.kind {
    |           ^^^^^^^^^ pattern `ExprKind::FormatArgs(_)` not covered
note: `ExprKind` defined here
note: `ExprKind` defined here
   --> /rustc/225ab3814cb608b5f14923fe372d5713e79a2b16/compiler/rustc_ast/src/ast.rs:1364:1
   ::: /rustc/225ab3814cb608b5f14923fe372d5713e79a2b16/compiler/rustc_ast/src/ast.rs:1504:5
    = note: not covered
    = note: the matched value is of type `ExprKind`
    = note: the matched value is of type `ExprKind`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
510 ~         | ast::ExprKind::Underscore => false,
510 ~         | ast::ExprKind::Underscore => false,
511 ~         ExprKind::FormatArgs(_) => todo!(),


error[E0004]: non-exhaustive patterns: `ExprKind::FormatArgs(_)` not covered
   |
   |
68 |     let expr_rw = match expr.kind {
   |                         ^^^^^^^^^ pattern `ExprKind::FormatArgs(_)` not covered
note: `ExprKind` defined here
note: `ExprKind` defined here
  --> /rustc/225ab3814cb608b5f14923fe372d5713e79a2b16/compiler/rustc_ast/src/ast.rs:1364:1
  ::: /rustc/225ab3814cb608b5f14923fe372d5713e79a2b16/compiler/rustc_ast/src/ast.rs:1504:5
   = note: not covered
   = note: the matched value is of type `ExprKind`
   = note: the matched value is of type `ExprKind`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
403~         ast::ExprKind::Err => None,
403~         ast::ExprKind::Err => None,
404~         ExprKind::FormatArgs(_) => todo!(),

For more information about this error, try `rustc --explain E0004`.
[RUSTC-TIMING] rustfmt_nightly test:false 3.770
error: could not compile `rustfmt-nightly` due to 2 previous errors
```

</details>

m-ou-se committed:
Assume there are no macros left in ast lowering.

m-ou-se committed:
Add FIXME comments about asm and format_args ast_pretty.

m-ou-se committed:
Update comments in rustc_ast_lowering/src/format.rs.

m-ou-se committed:
Add note on optimization in format args ast lowering.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit cd8ec6c787e147667e3bf43145873dd8e27755d0 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk cd8ec6c787e147667e3bf43145873dd8e27755d0 -->
<!-- homu: {"type":"Approved","sha":"cd8ec6c787e147667e3bf43145873dd8e27755d0","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

est31 commented on compiler/rustc_ast_lowering/src/format.rs:
```diff
@@ -223,6 +223,9 @@ fn expand_format_args<'hir>(
     // we don't do this, because an ArgumentV1 cannot be kept across yield points.
+    //
+    // This is an optimization, speeding up compilation about 1-2% in some cases.
+    // See https://perf.rust-lang.org/compare.html?start=5dbee4d3a6728eb4530fb66c9775834438ecec74&end=e36affffe97378a0027b4bcfbb18d27356164ed0&stat=instructions:u
```
I'm not sure how much of a permalink that is. Older perf results might get inaccessible in the future. I think it would be better to link to the github comment with the perf results instead, that contains a summary that is preserved, plus the discussion which lets you derive key facts from the context even if the perf run link itself is gone.

m-ou-se replied:
The entire summary is basically "speeding up compilation about 1-2% in some cases", which is included here.

Even if the link dies, the commit hashes will still be available, so it'd be easy to run the benchmarks again.

est31 replied:
I've checked perf runs of older PRs, many older ones are gone.

Gone (2017): https://github.com/rust-lang/rust/pull/45319#issuecomment-336907788

Gone (2019): https://github.com/rust-lang/rust/pull/60630#issuecomment-490561552

Still there (2020): https://github.com/rust-lang/rust/pull/78323#issuecomment-716140911

I'd prefer a link like: https://github.com/rust-lang/rust/pull/106745#issuecomment-1380268745

As for re-running benchmarks, I'm not that sure that it's this easy. In the time windows we are talking about (3-4 years), MSRV support of the benchmark suite, docker CI image and glibc version requirements start to play a role. It's probably possible but I doubt it is possible with the press of a button.

m-ou-se replied:
Updated the link to point to a github comment from the timer bot.

bors commented:
:hourglass: Testing commit cd8ec6c787e147667e3bf43145873dd8e27755d0 with merge c6192e5eae5cb6155b2d7c3b06b44d7f3d8b387e...
<!-- homu: {"type":"BuildStarted","head_sha":"cd8ec6c787e147667e3bf43145873dd8e27755d0","merge_sha":"c6192e5eae5cb6155b2d7c3b06b44d7f3d8b387e"} -->

m-ou-se committed:
Change not-so-permanent link to a more permanent link.

m-ou-se commented:
Looks like dist-x86_64-linux-alt got stuck. :(

m-ou-se commented:
@bors retry r=oli-obk

bors commented:
:pushpin: Commit 9f2ba59647b2cc70ce9467f2731e707459bf388f has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 9f2ba59647b2cc70ce9467f2731e707459bf388f -->
<!-- homu: {"type":"Approved","sha":"9f2ba59647b2cc70ce9467f2731e707459bf388f","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3913681300/jobs/6692798267) [(plain)](https://github.com/rust-lang-ci/rust/commit/c6192e5eae5cb6155b2d7c3b06b44d7f3d8b387e/checks/10640662077/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

SmiteWindows commented on compiler/rustc_ast_pretty/src/pprust/state/expr.rs:
```diff
@@ -627,3 +643,91 @@ impl<'a> State<'a> {
+                    }
+                }
+                if flags >> (rustc_parse_format::FlagDebugLowerHex as usize) & 1 != 0 {
+                    template.push('X');
```
template.push('x');

m-ou-se replied:
Great catch. Thanks!

SmiteWindows commented on compiler/rustc_ast_pretty/src/pprust/state/expr.rs:
```diff
@@ -627,3 +643,91 @@ impl<'a> State<'a> {
+                    template.push('X');
+                }
+                if flags >> (rustc_parse_format::FlagDebugUpperHex as usize) & 1 != 0 {
+                    template.push('x');
```
template.push('X');

m-ou-se committed:
Fix ast_pretty for format_args for {:x?} and {:X?}.

They were accidentally swapped.

m-ou-se commented:
@bors r=oli-obk

bors commented:
:pushpin: Commit 7b077307fd156275001ce5ca7a53e250eabcb0b6 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 7b077307fd156275001ce5ca7a53e250eabcb0b6 -->
<!-- homu: {"type":"Approved","sha":"7b077307fd156275001ce5ca7a53e250eabcb0b6","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 7b077307fd156275001ce5ca7a53e250eabcb0b6 with merge 099c1cf5e1ee0bfd829b97a064591cf838b62da1...
<!-- homu: {"type":"BuildStarted","head_sha":"7b077307fd156275001ce5ca7a53e250eabcb0b6","merge_sha":"099c1cf5e1ee0bfd829b97a064591cf838b62da1"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3921483753/jobs/6705297450) [(plain)](https://github.com/rust-lang-ci/rust/commit/099c1cf5e1ee0bfd829b97a064591cf838b62da1/checks/10655589396/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

matthiaskrgr commented:
@bors retry timeout

matthiaskrgr commented:
Hmm, this was `dist-x86_64-linux-alt` again, two times in a row, so maybe there is actually something weird going on and its hanging somewhere..
@bors r-

m-ou-se commented:
Oh interesting. Both times, the last log line was:

```
Documenting rustc_driver v0.0.0 (/checkout/compiler/rustc_driver)
```

If that's because of a bug in my code (quite possible), then it's surprising it only happens in dist-x86_64-linux-alt.

m-ou-se commented:
I can reproduce it locally.

m-ou-se commented:
It only happens when `rust.parallel-compiler = true` is enabled, which explains why it only happens in the -alt builder.

It gets stuck in `rustc_trait_selection` somehow. Here's a backtrace: [backtrace.txt](https://github.com/rust-lang/rust/files/10420013/backtrace.txt)

No clue what is happening.

m-ou-se commented:
@the8472 mentions that https://github.com/rust-lang/rust/pull/104754 is also stuck on the same issue.

m-ou-se commented:
Rustdoc also gets stuck at documenting `rustc_middle` at stage 1: `./x.py doc --stage 1 compiler/rustc_middle`

Compiling `rustc_middle` however works fine. So this seems to be specific to rustdoc, somehow.

oli-obk commented:
The root obligation that causes all the sadness is `<ty::fold::BottomUpFolder<'_, _, _, _> as std::marker::Send>`

which requires proving

```
Obligation(predicate=Binder(TraitPredicate(<ty::context::TyCtxt<'_> as std::marker::Send>, Polarity:Positive), []), depth=1)
Obligation(predicate=Binder(TraitPredicate(<_ as std::marker::Send>, polarity:Positive), []), depth=1)
Obligation(predicate=Binder(TraitPredicate(<_ as std::marker::Send>, polarity:Positive), []), depth=1)
Obligation(predicate=Binder(TraitPredicate(<_ as std::marker::Send>, polarity:Positive), []), depth=1)
Obligation(predicate=Binder(TraitPredicate(<ty::fold::BottomUpFolder<'_, _, _, _> as std::marker::Send>, polarity:Positive), []), depth=1)]
```

`<ty::context::TyCtxt<'_> as std::marker::Send>` gets proven sometimes, but is very expensive to prove. I wonder why it isn't cached.

I can't really tell where things go wrong, but it seems fishy that `BottomUpFolder as Send` causes this same obligation to show up in the nested obligations... Even if cycle detection or caching should catch this somewhere.

It may be there is no actual infinite recursion or infinite loop happening here, just that rustdoc tries to prove every blanket impl for every type, which turns out being very expensive once `Send` and `Sync` enter the game (which happens with parallel compiler). The main problem is likely the lack of caching, since afaict `TyCtxt: Send` gets proven a lot all over again.

Not sure what to do here.

m-ou-se commented:
Can we unblock this PR without solving that issue for now?

oli-obk commented:
We can disable that rustdoc feature under parallel-compiler.

m-ou-se commented:
> We can disable that rustdoc feature under parallel-compiler.

How?

oli-obk commented:
Adding `if cfg!(parallel_compiler) { return vec![]; }` to https://github.com/rust-lang/rust/blob/b8f9cb345ab1401f2fbd14cc23f64dda9dd2314e/src/librustdoc/clean/blanket_impl.rs#L15 should do the trick

m-ou-se commented:
I tried that, but that doesn't work. It `cfg(parallel_compiler)` doesn't seem to be defined for this crate.

oli-obk commented:
Uh.... I was sure i saw `#[cfg(parallel_compiler)]` somewhere in rustdoc, so I thought that would work...

@GuillaumeGomez any ideas how to disable parts of rustdoc in parallel compiler mode?

m-ou-se commented:
I assume that even if we disable that part, it'd fail many of its tests.

m-ou-se committed:
Work around issue 106930.

m-ou-se commented:
Added a workaround for #106930: an explicit Send and Sync impl for FormatArguments (only for cfg(parallel_compiler)), to stop rustdoc from recursing too deep into that when trying to prove Send/Sync.

m-ou-se commented:
Oli said this workaround is acceptable for landing this PR while we figure out the underlying issue.

@bors r=oli-obk

bors commented:
:pushpin: Commit db731e42b3c139587cd7a87acaa92cd82ab8a11c has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk db731e42b3c139587cd7a87acaa92cd82ab8a11c -->
<!-- homu: {"type":"Approved","sha":"db731e42b3c139587cd7a87acaa92cd82ab8a11c","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit db731e42b3c139587cd7a87acaa92cd82ab8a11c with merge 3e977638728922d3a6cc7bea34a2fdb8ae97f7c0...
<!-- homu: {"type":"BuildStarted","head_sha":"db731e42b3c139587cd7a87acaa92cd82ab8a11c","merge_sha":"3e977638728922d3a6cc7bea34a2fdb8ae97f7c0"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4015251396/jobs/6900416018)
Approved by: oli-obk
Pushing 3e977638728922d3a6cc7bea34a2fdb8ae97f7c0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4015251396/jobs/6900416018"},"merge_sha":"3e977638728922d3a6cc7bea34a2fdb8ae97f7c0"} -->

m-ou-se commented:
:tada:!

rust-timer commented:
Finished benchmarking commit (3e977638728922d3a6cc7bea34a2fdb8ae97f7c0): [comparison URL](https://perf.rust-lang.org/compare.html?start=40fda7b3fe2b10c6e1a0568b59516f5e7f381886&end=3e977638728922d3a6cc7bea34a2fdb8ae97f7c0&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=40fda7b3fe2b10c6e1a0568b59516f5e7f381886&end=3e977638728922d3a6cc7bea34a2fdb8ae97f7c0&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.2%, 0.8%]   | 4     |
| Regressions ❌ <br /> (secondary)  | 0.8%  | [0.3%, 1.3%]   | 3     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-2.3%, -0.2%] | 51    |
| Improvements ✅ <br /> (secondary) | -0.7% | [-0.8%, -0.7%] | 6     |
| All ❌✅ (primary)                 | -0.6% | [-2.3%, 0.8%]  | 55    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=40fda7b3fe2b10c6e1a0568b59516f5e7f381886&end=3e977638728922d3a6cc7bea34a2fdb8ae97f7c0&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4%  | [0.4%, 0.4%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 3.3%  | [2.0%, 7.9%]   | 7     |
| Improvements ✅ <br /> (primary)   | -2.9% | [-3.8%, -2.1%] | 4     |
| Improvements ✅ <br /> (secondary) | -3.1% | [-3.9%, -1.8%] | 3     |
| All ❌✅ (primary)                 | -2.2% | [-3.8%, 0.4%]  | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=40fda7b3fe2b10c6e1a0568b59516f5e7f381886&end=3e977638728922d3a6cc7bea34a2fdb8ae97f7c0&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.6%  | [1.5%, 1.8%]   | 9     |
| Regressions ❌ <br /> (secondary)  | 3.0%  | [1.5%, 4.9%]   | 18    |
| Improvements ✅ <br /> (primary)   | -2.2% | [-3.1%, -1.3%] | 5     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.3%  | [-3.1%, 1.8%]  | 14    |
</details>

nnethercote commented:
The perf improvements greatly outweigh the regressions.

@rustbot label: +perf-regression-triaged
